Formula/kubent.rb

```diff
@@ -0,0 +1,24 @@
+class Kubent < Formula
+  desc "Easily check your clusters for use of deprecated APIs"
+  homepage "https://github.com/doitintl/kube-no-trouble"
+  url "https://github.com/doitintl/kube-no-trouble.git",
+      tag:      "0.7.0",
+      revision: "d1bb4e5fd6550b533b2013671aa8419d923ee042"
+  license "MIT"
+  head "https://github.com/doitintl/kube-no-trouble.git", branch: "master"
+
+  depends_on "go" => :build
+
+  def install
+    ldflags = %W[
+      -s -w
+      -X main.version=#{version}
+      -X main.gitSha=#{Utils.git_head}
+    ]
+    system "go", "build", *std_go_args(ldflags: ldflags), "cmd/kubent/main.go"
+  end
+
+  test do
+    pipe_output("#{bin}/kubent -v", "INF version #{version}")
```

Comment:
can we check for some output here?

Reply:
@SMillerDev I may missunderstood documentation, but I understand L22 does "Check if "INF version #{version}" is included in the output of `#{bin}/kubent -v`", which would be an output check. Am I wrong?

Reply:
Yes, the current check is equivalent to running `echo "INF version #{version}" | kubent -v`. What you want would be
```suggestion
    assert_match "INF version #{version}", shell_output("#{bin}/kubent -v")
```

We need a test that exercises the some of the functionality of the app though. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
@SMillerDev I know that this test is a bad test, but the cookbook says better a bad test than no test. I can't imagine a better test. The issue is, that's a scanner for scanning Kubernetes clusters. So in order to test it, one would need to provide a cluster. Think about a client for a backend, where you can't test the client without having a backend it can connect to. Does that make sense?

I tried the `assert_match` before, but didn't got it working. I added it again, and here's what I get:

```
$ brew test Formula/kubent.rb
==> Testing kubent
==> /opt/homebrew/Cellar/kubent/0.7.0/bin/kubent -v
2:53PM INF version 0.7.0 (git sha d1bb4e5fd6550b533b2013671aa8419d923ee042)
Error: kubent: failed
An exception occurred within a child process:
  Minitest::Assertion: Expected /INF\ version\ 0\.7\.0/ to match "".
/opt/homebrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.17.0/lib/minitest/assertions.rb:183:in `assert'
/opt/homebrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.17.0/lib/minitest/assertions.rb:295:in `assert_match'
/Users/tdeutsch/development/github/tuxpeople/homebrew-core/Formula/kubent.rb:22:in `block in <class:Kubent>'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2301:in `block (3 levels) in run_test'
/opt/homebrew/Homebrew/Library/Homebrew/utils.rb:605:in `with_env'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2300:in `block (2 levels) in run_test'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:968:in `with_logging'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2299:in `block in run_test'
/opt/homebrew/Homebrew/Library/Homebrew/mktemp.rb:77:in `block in run'
/opt/homebrew/Homebrew/Library/Homebrew/mktemp.rb:77:in `chdir'
/opt/homebrew/Homebrew/Library/Homebrew/mktemp.rb:77:in `run'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2577:in `mktemp'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2293:in `run_test'
/opt/homebrew/Homebrew/Library/Homebrew/test.rb:45:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/opt/homebrew/Homebrew/Library/Homebrew/test.rb:50:in `<main>'
```

For me it looks like some sort of escaping issue.

Reply:
I just reread the last few bullet points of your feedback, let me check...

Reply:
@SMillerDev this is what happens, when I use the tool without providing connection details:

```
$ kubent
3:00PM INF >>> Kube No Trouble `kubent` <<<
3:00PM INF version 0.7.0 (git sha d1bb4e5fd6550b533b2013671aa8419d923ee042)
3:00PM INF Initializing collectors and retrieving data
3:00PM ERR Failed to initialize collector: <nil> error="failed to assemble client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable"
3:00PM ERR Failed to initialize collector: <nil> error="failed to assemble client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable"
3:00PM INF Loaded ruleset name=custom.rego.tmpl
3:00PM INF Loaded ruleset name=deprecated-1-16.rego
3:00PM INF Loaded ruleset name=deprecated-1-22.rego
3:00PM INF Loaded ruleset name=deprecated-1-25.rego
3:00PM INF Loaded ruleset name=deprecated-1-26.rego
3:00PM INF Loaded ruleset name=deprecated-future.rego
```

Therefore, I changed my test to this:
```
  test do
    assert_match "no configuration has been provided", shell_output("#{bin}/kubent")
  end
```

However, it doesn't work and I'm not sure why:
```
$ brew test Formula/kubent.rb
==> Testing kubent
==> /opt/homebrew/Cellar/kubent/0.7.0/bin/kubent
3:01PM INF >>> Kube No Trouble `kubent` <<<
3:01PM INF version 0.7.0 (git sha d1bb4e5fd6550b533b2013671aa8419d923ee042)
3:01PM INF Initializing collectors and retrieving data
3:01PM ERR Failed to initialize collector: <nil> error="failed to assemble client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable"
3:01PM ERR Failed to initialize collector: <nil> error="failed to assemble client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable"
3:01PM INF Loaded ruleset name=custom.rego.tmpl
3:01PM INF Loaded ruleset name=deprecated-1-16.rego
3:01PM INF Loaded ruleset name=deprecated-1-22.rego
3:01PM INF Loaded ruleset name=deprecated-1-25.rego
3:01PM INF Loaded ruleset name=deprecated-1-26.rego
3:01PM INF Loaded ruleset name=deprecated-future.rego
Error: kubent: failed
An exception occurred within a child process:
  Minitest::Assertion: Expected /no\ configuration\ has\ been\ provided/ to match "".
/opt/homebrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.17.0/lib/minitest/assertions.rb:183:in `assert'
/opt/homebrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.17.0/lib/minitest/assertions.rb:295:in `assert_match'
/Users/tdeutsch/development/github/tuxpeople/homebrew-core/Formula/kubent.rb:22:in `block in <class:Kubent>'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2301:in `block (3 levels) in run_test'
/opt/homebrew/Homebrew/Library/Homebrew/utils.rb:605:in `with_env'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2300:in `block (2 levels) in run_test'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:968:in `with_logging'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2299:in `block in run_test'
/opt/homebrew/Homebrew/Library/Homebrew/mktemp.rb:77:in `block in run'
/opt/homebrew/Homebrew/Library/Homebrew/mktemp.rb:77:in `chdir'
/opt/homebrew/Homebrew/Library/Homebrew/mktemp.rb:77:in `run'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2577:in `mktemp'
/opt/homebrew/Homebrew/Library/Homebrew/formula.rb:2293:in `run_test'
/opt/homebrew/Homebrew/Library/Homebrew/test.rb:45:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/opt/homebrew/Homebrew/Library/Homebrew/test.rb:50:in `<main>'
```

I'd appreciate it if you could provide me with a hint.

Reply:
Does it write to stdout or stderr? Because `shell_output` only captures stdout.

Reply:
It's just a shell that it gets executed in so `2&>` should work IIRC

Reply:
It's `2>&1` and I fixed it. Does it look good for you now?